sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/BatchRequestForRecommendationAI.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.ml;
+
+import com.google.api.client.json.GenericJson;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.state.BagState;
+import org.apache.beam.sdk.state.StateSpec;
+import org.apache.beam.sdk.state.StateSpecs;
+import org.apache.beam.sdk.state.TimeDomain;
+import org.apache.beam.sdk.state.Timer;
+import org.apache.beam.sdk.state.TimerSpec;
+import org.apache.beam.sdk.state.TimerSpecs;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.KV;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * DoFn that batches up GenericJson to reduce the number of requests made to the Recommendations
+ * API.
+ */
+class BatchRequestForRecommendationAI
+    extends DoFn<KV<String, GenericJson>, KV<String, Iterable<GenericJson>>> {
+  private static final Logger LOG = LoggerFactory.getLogger(BatchRequestForRecommendationAI.class);
+
+  private final Counter numberOfRowsBagged =
+      Metrics.counter(BatchRequestForRecommendationAI.class, "numberOfRowsBagged");
+
+  private final Integer maxBatchSize;
+
+  @StateId("elementsBag")
+  private final StateSpec<BagState<KV<String, GenericJson>>> elementsBag = StateSpecs.bag();
+
+  @TimerId("eventTimer")
+  private final TimerSpec eventTimer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+  /**
+   * Constructs the batching DoFn.
+   *
+   * @param maxBatchSize Desired batch size in bytes.
+   */
+  public BatchRequestForRecommendationAI(Integer maxBatchSize) {
+    this.maxBatchSize = maxBatchSize;
+  }
+
+  @ProcessElement
+  public void process(
+      @Element KV<String, GenericJson> element,
+      @StateId("elementsBag") BagState<KV<String, GenericJson>> elementsBag,
+      @TimerId("eventTimer") Timer eventTimer,
+      BoundedWindow w) {
+    elementsBag.add(element);
+    eventTimer.set(w.maxTimestamp());
```

Comment:
Consider a streaming pipeline that streams elements in the global window (a simple pipeline that does not use windowing is fairly commont) - in this case, this timer would never be triggered, right?

I recommend that instead of using this transform, you can use [GroupIntoBatches](https://beam.apache.org/releases/javadoc/2.27.0/org/apache/beam/sdk/transforms/GroupIntoBatches.html), which implements this functionality, and receives other updates (like auto-sharded keys)

Reply:
Good point, makes sense to re-use this functionality (wasn't aware of it yet). Will readjust the transforms and delete the unnecessary ones.